Protocol Upgrades

kdelorey opened the pull request:
This PR adds support for Envoy's protocol upgrades, with WebSockets being supported. This allows you to configure the use of protocol upgrades as part of the HTTP Connection Manager on the Gateway or on route options. The configuration is the same for both, see examples below.

HCM Example:
```yaml
httpGateway:
  options:
    httpConnectionManagerSettings:
      upgrades:
       - websocket:
            enabled: false
```

Route Options Example:
```yaml
- matchers:
    - prefix: /foo
   routeAction:
     single:
       upstream:
         name: foo
         namespace: gloo-system
    options:
      upgrades:
      - websocket: {}
```

Envoy supports custom filtering as part of the upgrade configuration. This approach leaves the door open for support there if it is deemed required.
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/1132

kdelorey commented:
Something to note is that WebSockets seemed to be turned on [here](https://github.com/solo-io/gloo/blob/edf4a9d39f81eceabd6bc1061df7c52259ca9f18/projects/gloo/pkg/translator/http_filters.go#L38-L40).

I'm not sure what should be done about it? Options include:
- Breaking change, remove the referenced code and require users to enable it from the HCM or route.
- Do nothing: maybe document that WebSockets are turned on by default and you have to explicitly turn them off on the HCM.
- Move the offending code to the HCM plugin. If left unset in the configuration, it will enable WebSockets by default keeping the same behavior.

kdelorey commented:
Are there any thoughts on the option name? I was considering renaming it to just `upgrades` on the Route Options.

kdorosh commented on projects/gloo/api/v1/options/hcm/hcm.proto:
```diff
@@ -52,4 +53,7 @@ message HttpConnectionManagerSettings {
 
     SetCurrentClientCertDetails set_current_client_cert_details = 19;
+
+    // Route configuration for protocol upgrade requests.
```
```suggestion
    // HttpConnectionManager configuration for protocol upgrade requests.
```

kdorosh commented on projects/gloo/api/v1/options/upgrade/upgrade.proto:
```diff
@@ -0,0 +1,22 @@
+option (gogoproto.equal_all) = true;
+
+message UpgradeConfig {
+    message UpgradeSpec {
```
can we rename this message to something more specific? if we added another upgrade spec type in the future we would want each to have specific names

kdelorey replied:
Yeah, I also had the same thought. Maybe just `ProtocolUpgradeConfig` and `ProtocolUpgradeSpec`?

kdorosh replied:
I was thinking more along the lines of renaming `UpgradeSpec` to `WebsocketUpgrade`

kdorosh replied:
I like the other rename as well

kdelorey replied:
Oh, the reason why I made it more general is to allow the ability to add other protocols and keeps things DRY. The options would be exactly the same (and can include [custom filters](https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/network/http_connection_manager/v2/http_connection_manager.proto#config-filter-network-http-connection-manager-v2-httpconnectionmanager-upgradeconfig)).

That being said, I'm not even sure what other protocols upgrades even exist.

The interface, if some other protocol would exist, would look like:
```yaml
upgradeConfigs:
- websocket:
    enabled: false
- protocolX:
    enabled: false
```

Do you still feel like it should be specific to websockets?

kdorosh replied:
This is what you have currrently:
```proto
message UpgradeConfig {
    message UpgradeSpec {
        // Whether the upgrade should be enabled. `true` by default.
        google.protobuf.BoolValue enabled = 1;
    }

    oneof upgrade_type {
        // Specification for websocket upgrade requests.
        UpgradeSpec websocket = 1;
    }
}
```

Now let's say you want to add another upgrade type. You would have to do something gross, like:
```proto
message UpgradeConfig {
    message UpgradeSpec {
        // Whether the websocket upgrade should be enabled. `true` by default.
        google.protobuf.BoolValue enabled = 1;
    }

    // without the 2 suffix or something to differentiate it, we 
    // would have a name collision with the other upgrade spec
    message UpgradeSpec2 {
        // Whether the upgrade should be enabled. `true` by default.
        google.protobuf.BoolValue enabled = 1;
    }

    oneof upgrade_type {
        // Specification for websocket upgrade requests.
        UpgradeSpec websocket = 1;
        // Specification for another upgrade requests.
        UpgradeSpec2 another = 2;
    }
}
```

Instead I'd like to nip it now with a better name. I'm proposing:
```proto
message UpgradeConfig {
    message WebsocketUpgrade {
        // Whether the websocket upgrade should be enabled. `true` by default.
        google.protobuf.BoolValue enabled = 1;
    }

    message AnotherUpgrade {
        // Whether the upgrade should be enabled. `true` by default.
        google.protobuf.BoolValue enabled = 1;
    }

    oneof upgrade_type {
        // Specification for websocket upgrade requests.
        WebsocketUpgrade websocket = 1;
        // Specification for another upgrade requests.
        AnotherUpgrade another = 2;
    }
}
```

kdorosh replied:
The alternative is to move `UpgradeSpec` to other proto files and give them different packages, but I prefer more specific names

kdelorey replied:
Oh I see, I guess I just misunderstood how I might do this in protobuf. I thought I could reuse the spec like:

```
oneof upgrade_type {
   UpgradeSpec websocket = 1;
   UpgradeSpec another = 2;
}
```
Anyways, I'll fix it to be more specific.

Thanks for the guidance!

kdorosh replied:
Ah, after looking more closely at your proposed solution and the underlying envoy config, I think your solution is sufficient and mine is overkill.

If we ever wanted to do something fancy with the upgrade config we could just add the specific  `UpgradeSpec<blah>` there, but the current config should cover most cases, as the underlying envoy config just has a `upgrade_type`, `enabled`, and `filters` (`filters` not currently exposed, but non-breaking to add if ever desired).

I was just being a bit paranoid.

Thanks for your work and attention to defail @kdelorey !

kdorosh commented on projects/gloo/pkg/plugins/basicroute/plugin.go:
```diff
@@ -126,6 +130,39 @@ func applyHostRewrite(in *v1.Route, out *envoyroute.Route) error {
+	for i, config := range upgrades {
+		switch upgradeType := config.GetUpgradeType().(type) {
+		default:
+			return errors.Errorf("unimplemented upgrade type: %T", upgradeType)
```
nit: I think the default case at the end is more idiomatic

kdelorey replied:
Agreed. I was sure that I saw this style used elsewhere in the gloo codebase and I just did it the same way. I can't an example of this at the moment, but I have no issues with changing this.

kdelorey replied:
I figured out that I saw the leading `default` case in the `applyHostRewrite` function just above this one method. I've taken the liberty in modifying it so it follows the same style.

kdorosh commented on projects/gloo/api/v1/options/upgrade/upgrade.proto:
```diff
@@ -0,0 +1,22 @@
+message UpgradeConfig {
+    message UpgradeSpec {
+        // Whether the upgrade should be enabled. `true` by default.
+        google.protobuf.BoolValue enabled = 1;
```
I think this defaults to false

kdelorey replied:
Yeah it does, but I guess I was documenting the behavior of Envoy here. If you were to just do: `websocket: {}` it would enable WebSockets. But it's really there so you can do:
```
websocket:
  enabled: false
```

This allows you to maybe turn on WebSockets in the HCM and turn them off for a specific route.

kdorosh replied:
I guess, more accurately, the upgrade option defaults to false but in the code we set it to true already

kdorosh commented on projects/gloo/pkg/plugins/hcm/plugin.go:
```diff
@@ -112,6 +112,26 @@ func copyCoreHcmSettings(cfg *envoyhttp.HttpConnectionManager, hcmSettings *hcm.
+					UpgradeType: "websocket",
+					Enabled:     config.GetWebsocket().GetEnabled(),
+				}
+			}
```
we should add the default case here too to ensure invalid upgrade types return in errors

kdorosh commented:
> Are there any thoughts on the option name? I was considering renaming it to just upgrades on the Route Options.

I think we should be consistent, and use `upgrade` or `upgradeConfig` in both places. I slightly prefer the former for brevity

kdorosh commented:
With respect to  https://github.com/solo-io/gloo/pull/1784#issuecomment-559599192

I think we should move the offending code to the plugin, make the default true (in the proto as well), and document that you have to turn it off.

We shouldn't make a breaking change here.

kdorosh commented on projects/gloo/pkg/plugins/basicroute/plugin.go:
```diff
@@ -126,6 +130,39 @@ func applyHostRewrite(in *v1.Route, out *envoyroute.Route) error {
+			return errors.Errorf("unimplemented upgrade type: %T", upgradeType)
+		case *upgrade.UpgradeConfig_Websocket:
+			routeAction.Route.UpgradeConfigs[i] = &envoyroute.RouteAction_UpgradeConfig{
+				UpgradeType: "websocket",
```
we should use a variable for this, https://github.com/solo-io/gloo/blob/edf4a9d39f81eceabd6bc1061df7c52259ca9f18/projects/gloo/pkg/translator/http_filters.go#L22

yuval-k commented:
@kdelorey note that gloo enables `websocket` upgrades by default.
see here: 
https://github.com/solo-io/gloo/blob/edf4a9d39f81eceabd6bc1061df7c52259ca9f18/projects/gloo/pkg/translator/http_filters.go#L39

in your use case, do you need it disabled by default, and enabled only for a specific route?
if so, I recommend adding an option to disable the default upgrade config as well

kdelorey commented:
> 
> 
> @kdelorey note that gloo enables `websocket` upgrades by default.
> see here:
> 
> https://github.com/solo-io/gloo/blob/edf4a9d39f81eceabd6bc1061df7c52259ca9f18/projects/gloo/pkg/translator/http_filters.go#L39
> 
> in your use case, do you need it disabled by default, and enabled only for a specific route?
> if so, I recommend adding an option to disable the default upgrade config as well

@yuval-k per @kdorosh's [comment](https://github.com/solo-io/gloo/pull/1784#issuecomment-561712468), he suggested my 3rd option which was to move the enabling to the HCM plugin which would keep the same behavior. Are you opposed to that?

With this PR you can disable WebSockets with the following HCM configuration:
```yaml
upgradeConfigs:
- websocket:
    enabled: false
```

Honestly, my use case was me attempting to figure out my WebSocket issues on my deployment. I ended up figuring it out; it ended up being related to session affinity. Regardless I saw this PR through as it seems like a nice to have. It would allow you to sanitize protocol upgrades only to specific routes if desired.

yuval-k commented:
@kdelorey  makes sense - as long as backwards compatibility is maintained, this sounds good to me.

kdelorey commented:
I think I've addressed everything we talked about @kdorosh and I'm ready for a formal review. As a summary of the changes:
- Merged `feature-rc1` into my branch and resolved conflicts and errors.
- Renamed the option on the HCM and Route Options to `upgrades` (from `upgradeConfigs`).
- Renamed the proto messages to `ProtocolUpgradeConfig` and `ProtocolUpgradeSpec` respectively.
- The core HCM translation can now report errors and I added the default case to upgrade type `switch`.
- Moved default enabling of WebSockets from the `http_filters.go` to the HCM plugin. Added unit test to ensure they are enabled by default within the config translation.

Thanks!

yuval-k commented:
/test

kdelorey commented:
@yuval-k will do. I can't get to it atm, but I'll be able to take care of it later today. Thanks for taking a look!

kdelorey commented:
@yuval-k should be good to go now, I'm guessing I can't run

/test

myself, but I can try ¯\_(ツ)_/¯

yuval-k commented:
/test

yuval-k commented:
looking good, thanks @kdelorey !

yuval-k commented:
> @yuval-k should be good to go now, I'm guessing I can't run
> 
> /test
> 
> myself, but I can try ¯_(ツ)_/¯

yes, the bot checks that you are in the org; that's our `use our CI for crypto-mining` mitigation

yuval-k commented:
@kdelorey 
i see a test failure in CI:
```

Step #11 - "test": ------------------------------
Step #11 - "test": • Failure [0.003 seconds]
Step #11 - "test": HttpFilters [It] should have websocket upgrade config 
Step #11 - "test": /workspace/gopath/src/github.com/solo-io/gloo/projects/gloo/pkg/translator/http_filters_test.go:12
Step #11 - "test": 
Step #11 - "test":   Expected
Step #11 - "test":       <[]*envoy_config_filter_network_http_connection_manager_v2.HttpConnectionManager_UpgradeConfig | len:0, cap:0>: nil
Step #11 - "test":   to have length 1
Step #11 - "test": 
Step #11 - "test":   /workspace/gopath/src/github.com/solo-io/gloo/projects/gloo/pkg/translator/http_filters_test.go:14
Step #11 - "test": 
Step #11 - "test":   Full Stack Trace
Step #11 - "test":   github.com/solo-io/gloo/projects/gloo/pkg/translator_test.glob..func1.1()
Step #11 - "test":   	/workspace/gopath/src/github.com/solo-io/gloo/projects/gloo/pkg/translator/http_filters_test.go:14 +0x550
Step #11 - "test":   github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0xc00011e540, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
Step #11 - "test":   	/workspace/gopath/pkg/mod/github.com/onsi/ginkgo@v1.10.1/internal/leafnodes/runner.go:113 +0xe0
Step #11 - "test":   github.com/onsi/ginkgo/internal/leafnodes.(*runner).run(0xc00011e540, 0xc00041a768, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
Step #11 - "test":   	/workspace/gopath/pkg/mod/github.com/onsi/ginkgo@v1.10.1/internal/leafnodes/runner.go:64 +0x108
Step #11 - "test":   github.com/onsi/ginkgo/internal/leafnodes.(*ItNode).Run(0xc00000ee20, 0x3497080, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
Step #11 - "test":   	/workspace/gopath/pkg/mod/github.com/onsi/ginkgo@v1.10.1/internal/leafnodes/it_node.go:26 +0x88
Step #11 - "test":   github.com/onsi/ginkgo/internal/spec.(*Spec).runSample(0xc00012bc20, 0x0, 0x3497080, 0xc00041a740)
Step #11 - "test":   	/workspace/gopath/pkg/mod/github.com/onsi/ginkgo@v1.10.1/internal/spec/spec.go:215 +0x6e2
Step #11 - "test":   github.com/onsi/ginkgo/internal/spec.(*Spec).Run(0xc00012bc20, 0x3497080, 0xc00041a740)
Step #11 - "test":   	/workspace/gopath/pkg/mod/github.com/onsi/ginkgo@v1.10.1/internal/spec/spec.go:138 +0x168
Step #11 - "test":   github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec(0xc000568c80, 0xc00012bc20, 0x0)
Step #11 - "test":   	/workspace/gopath/pkg/mod/github.com/onsi/ginkgo@v1.10.1/internal/specrunner/spec_runner.go:200 +0x173
Step #11 - "test":   github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs(0xc000568c80, 0x1)
Step #11 - "test":   	/workspace/gopath/pkg/mod/github.com/onsi/ginkgo@v1.10.1/internal/specrunner/spec_runner.go:170 +0x217
Step #11 - "test":   github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run(0xc000568c80, 0x8)
Step #11 - "test":   	/workspace/gopath/pkg/mod/github.com/onsi/ginkgo@v1.10.1/internal/specrunner/spec_runner.go:66 +0x141
Step #11 - "test":   github.com/onsi/ginkgo/internal/suite.(*Suite).Run(0xc0002b7d10, 0x7f5ac04fd188, 0xc00005f300, 0x2f3d040, 0x10, 0xc0002f01b0, 0x1, 0x1, 0x350b440, 0xc00041a740, ...)
Step #11 - "test":   	/workspace/gopath/pkg/mod/github.com/onsi/ginkgo@v1.10.1/internal/suite/suite.go:62 +0x6a0
Step #11 - "test":   github.com/onsi/ginkgo.RunSpecsWithCustomReporters(0x3499160, 0xc00005f300, 0x2f3d040, 0x10, 0xc0000cded8, 0x1, 0x1, 0x20)
Step #11 - "test":   	/workspace/gopath/pkg/mod/github.com/onsi/ginkgo@v1.10.1/ginkgo_dsl.go:221 +0x339
Step #11 - "test":   github.com/onsi/ginkgo.RunSpecs(0x3499160, 0xc00005f300, 0x2f3d040, 0x10, 0x0)
Step #11 - "test":   	/workspace/gopath/pkg/mod/github.com/onsi/ginkgo@v1.10.1/ginkgo_dsl.go:202 +0x9a
Step #11 - "test":   github.com/solo-io/gloo/projects/gloo/pkg/translator_test.TestTranslator(0xc00005f300)
Step #11 - "test":   	/workspace/gopath/src/github.com/solo-io/gloo/projects/gloo/pkg/translator/translator_suite_test.go:15 +0x119
Step #11 - "test":   testing.tRunner(0xc00005f300, 0x30015e8)
Step #11 - "test":   	/usr/local/go/src/testing/testing.go:909 +0x19a
Step #11 - "test":   created by testing.(*T).Run
Step #11 - "test":   	/usr/local/go/src/testing/testing.go:960 +0x652
```

kdelorey commented:
Oh weird, I ran `ginkgo` locally to check all that and it passed. I'll take a look, thanks!

kdelorey commented:
I guess one question here is that the `http_filters_test.go` only has one test in it that tests to see if Websockets are enabled. I need to remove this test but if I remove there would be no tests at all. Should I just remove the whole file?

yuval-k commented:
I would say deleted it, as most of the tests in that package are in translator_test.go

kdelorey commented:
Alright, I removed the test file. I'm confident it'll work this time ;)

yuval-k commented:
/test
